File: algos_assignment/team_pratice/ACM ICPC/떡볶이파스타/9월_3주차/4949.py
while True:
    a = input()
    if a==".":
        break
    stack = []
    hi = 1
    for i in a:
        if i=='(' or i=='[':
            stack.append(i)

        elif i==')':
            if stack and stack[-1]=='(':
                stack.pop()
            else:
                hi = 0
                break

        elif i==']':
            if stack and stack[-1]=='[':
                stack.pop()
            else:
                hi = 0
                break

    print("yes" if hi and not(stack) else "no")
            

File: algos_assignment/2nd_grade/김민정_mzeong/week1/4949.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int main()
{
	string str;

	while (getline(cin, str))
	{
		stack<char> s;
		bool bal = true; // 참과 거짓 값만 가질 수 있는 변수

		if (str.size() == 1 && str[0] == '.')
			break;
		for (int i = 0; i < str.size(); i++)
		{
			if (str[i] == '(' || str[i] == '[')
				s.push(str[i]);
			else if (str[i] == ')')
			{
				if (!s.empty() && s.top() == '(')
					s.pop();
				else
				{
					bal = false;
					break;
				}
			}
			else if (str[i] == ']')
			{
				if (!s.empty() && s.top() == '[')
					s.pop();
				else 
				{
					bal = false;
					break;
				}
			}
		}
		if (s.empty() && bal)
			cout << "yes\n";
		else
			cout << "no\n";
	}
}
